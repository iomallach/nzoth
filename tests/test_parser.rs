mod helpers;
use std::cell::RefCell;

use helpers::SnapshotErrors;
use insta;
use nzoth::{parser::Parser, source::SourceFile};

#[test]
fn test_valid_let_declarations() {
    insta::glob!("resources/parser_cases/valid_let_declarations.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();
        assert_eq!(2, program.nodes.len());

        let output = format!("{:#?}", program);

        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_valid_identifier() {
    insta::glob!("resources/parser_cases/valid_identifiers.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let output = format!("{:#?}", program);
        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_valid_assignments() {
    insta::glob!("resources/parser_cases/valid_assignments.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        let program = parser.parse();

        let output = format!("{:#?}", program);
        insta::assert_snapshot!(output);
    })
}

#[test]
fn test_invalid_assignments() {
    insta::glob!("resources/parser_cases/invalid_assignments.nz", |path| {
        let contents = std::fs::read_to_string(path).expect("Expect path exists");
        let source = RefCell::new(SourceFile::new(
            0,
            path.file_name()
                .expect("Expect a file")
                .to_str()
                .expect("No fail")
                .to_string(),
            contents.as_str(),
        ));

        let mut parser = Parser::new(&source);
        _ = parser.parse();

        insta::assert_snapshot!(SnapshotErrors(parser.errors()));
    })
}
