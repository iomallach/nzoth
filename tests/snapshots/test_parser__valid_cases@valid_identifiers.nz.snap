---
source: tests/test_parser.rs
input_file: tests/resources/parser_cases/valid_identifiers.nz
---
Program {
    nodes: [
        FunctionDeclaration(
            FuncDeclaration {
                identifier: "main",
                paramemetrs: [],
                body: Block {
                    nodes: [
                        Expression(
                            ExpressionStatement {
                                span: Span {
                                    start: Position {
                                        offset: 18,
                                        line: 1,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 21,
                                        line: 1,
                                        column: 5,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                                expression: Identifier(
                                    "foo",
                                    Span {
                                        start: Position {
                                            offset: 18,
                                            line: 1,
                                            column: 2,
                                        },
                                        end: Position {
                                            offset: 21,
                                            line: 1,
                                            column: 5,
                                        },
                                        file: FileId(
                                            0,
                                        ),
                                    },
                                ),
                            },
                        ),
                        Expression(
                            ExpressionStatement {
                                span: Span {
                                    start: Position {
                                        offset: 25,
                                        line: 2,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 28,
                                        line: 2,
                                        column: 5,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                                expression: Identifier(
                                    "bar",
                                    Span {
                                        start: Position {
                                            offset: 25,
                                            line: 2,
                                            column: 2,
                                        },
                                        end: Position {
                                            offset: 28,
                                            line: 2,
                                            column: 5,
                                        },
                                        file: FileId(
                                            0,
                                        ),
                                    },
                                ),
                            },
                        ),
                    ],
                    last_expression: Some(
                        ExpressionStatement {
                            span: Span {
                                start: Position {
                                    offset: 32,
                                    line: 3,
                                    column: 2,
                                },
                                end: Position {
                                    offset: 35,
                                    line: 3,
                                    column: 5,
                                },
                                file: FileId(
                                    0,
                                ),
                            },
                            expression: Identifier(
                                "baz",
                                Span {
                                    start: Position {
                                        offset: 32,
                                        line: 3,
                                        column: 2,
                                    },
                                    end: Position {
                                        offset: 35,
                                        line: 3,
                                        column: 5,
                                    },
                                    file: FileId(
                                        0,
                                    ),
                                },
                            ),
                        },
                    ),
                    span: Span {
                        start: Position {
                            offset: 14,
                            line: 0,
                            column: 14,
                        },
                        end: Position {
                            offset: 37,
                            line: 4,
                            column: 1,
                        },
                        file: FileId(
                            0,
                        ),
                    },
                },
                return_type: None,
                span: Span {
                    start: Position {
                        offset: 0,
                        line: 0,
                        column: 0,
                    },
                    end: Position {
                        offset: 3,
                        line: 0,
                        column: 3,
                    },
                    file: FileId(
                        0,
                    ),
                },
            },
        ),
    ],
}
